projects/openapi-utilities/src/utilities/json-changelog.ts

```diff
@@ -0,0 +1,333 @@
+): ChangedNode['attributes'] {
+  const { label, excludeKeys } = options;
+  const diff = getDetailsDiff(change, excludeKeys);
+
+  const diffCount =
+    diff.changed.length + diff.removed.length + diff.added.length;
+
+  if (!diffCount) return [];
+
+  const results: ChangedNode['attributes'] = [];
+
+  // don't show children as added
+  if (!change.added) {
+    diff.added.forEach((added) =>
+      results.push({ key: added.key, after: added.value, before: undefined })
+    );
+  }
+  // don't show children as removed
+  if (!change.removed) {
+    diff.removed.forEach((removed) =>
+      results.push({
+        key: removed.key,
+        before: removed.value,
+        after: undefined,
+      })
+    );
+  }
+  diff.changed.forEach((changed) =>
+    results.push({
+      key: changed.key,
+      before: changed.before,
+      after: changed.after,
+    })
+  );
+
+  return results;
+}
+
+type ChangedNode = {
+  name: string;
+  change: 'added' | 'removed' | 'changed';
+  attributes: { key: string; before: any; after: any }[];
+};
+
+type OperationChangelog = ChangedNode & {
+  parameters: ChangedNode[];
+  requestBody?: ChangedNode;
+  responses: ChangedNode[];
+};
+
+type JsonChangelog = {
+  operations: OperationChangelog[];
+};
+
+export function jsonChangelog(
+  groupedChanges: ReturnType<typeof groupChangesAndRules>
+): JsonChangelog {
+  const results: JsonChangelog = { operations: [] };
+
+  const { changesByEndpoint, specification } = groupedChanges;
+  for (const [_, endpointChange] of changesByEndpoint) {
+    results.operations.push(getEndpointLogs(endpointChange));
+  }
+
+  return results;
+}
+
+function getEndpointLogs(
+  endpointChange: OpenApiEndpointChange
+): OperationChangelog {
+  const {
+    method,
+    path,
+    request,
+    responses,
+    cookieParameters,
+    change,
+    headers,
+    pathParameters,
+    queryParameters,
+  } = endpointChange;
+
+  const parameterChanges = [];
+  for (const [name, parameterChange] of queryParameters.changes) {
+    parameterChanges.push(getParameterLogs('query', name, parameterChange));
+  }
+
+  for (const [name, parameterChange] of cookieParameters.changes) {
+    parameterChanges.push(getParameterLogs('cookie', name, parameterChange));
+  }
+
+  for (const [name, parameterChange] of pathParameters.changes) {
+    parameterChanges.push(getParameterLogs('path', name, parameterChange));
+  }
+
+  for (const [name, parameterChange] of headers.changes) {
+    parameterChanges.push(getParameterLogs('header', name, parameterChange));
+  }
+
+  const responseChanges: ChangedNode[] = [];
+  for (const [key, response] of responses) {
+    responseChanges.push(getResponseChangeLogs(response, key));
+  }
+
+  return {
+    name: `${method} ${path}`,
+    change: getChange(change),
+    attributes: change
+      ? getDetailLogs(change, {
+          excludeKeys: ['pathPattern', 'method'],
+        })
+      : [],
+    parameters: parameterChanges,
+    requestBody:
```

Comment:
if !added !change !removed -- is this always null?

Reply:
Yeah it should be